Java/core/src/main/java/com/amazon/randomcutforest/store/IndexManager.java

```diff
@@ -30,54 +31,110 @@
+            freeIndexPointer = (capacity - 1);
+            occupied = new BitSet(capacity);
+        } else {
+            freeIndexPointer = -1;
+            occupied = bits;
+            for (int i = 0; i < capacity; i++) {
+                if (!occupied.get(i)) {
+                    freeIndexPointer++;
+                }
+            }
+            if (freeIndexPointer != capacity - 1) {
+                freeIndexes = new int[freeIndexPointer + 1];
+                int location = 0;
+                for (int i = capacity - 1; i >= 0; i--) {
+                    if (!occupied.get(i)) {
+                        freeIndexes[location++] = i;
+                    }
+                }
+            } else {
+                freeIndexes = new int[0];
+            }
         }
-        freeIndexPointer = capacity - 1;
-
     }
 
     /**
-     * Construct a new IndexManager with the given array of free indexes.
-     * 
-     * @param freeIndexes      An array of index values.
+     * Construct a new IndexManager with the given capacity, array of free indexes
+     * and a value corresponding to the number of free indices
+     *
+     * @param capacity         the size of the index
+     * @param freeIndexes      An array of unoccupied index values.
      * @param freeIndexPointer Entries in freeIndexes between 0 (inclusive) and
      *                         freeIndexPointer (inclusive) contain valid index
-     *                         values.
+     *                         values. if freeIndexPointer is larger than the length
+     *                         of the freeIndices array then the implcit guarantee
+     *                         is that location i greater or equal
+     *                         freeIndexes.length contains index value (capacity - i
+     *                         -1)
      */
-
-    public IndexManager(int[] freeIndexes, int freeIndexPointer) {
+    public IndexManager(int capacity, int[] freeIndexes, int freeIndexPointer) {
         checkNotNull(freeIndexes, "freeIndexes must not be null");
-        checkFreeIndexes(freeIndexes, freeIndexPointer);
 
-        this.capacity = freeIndexes.length;
+        this.capacity = capacity;
         this.freeIndexes = freeIndexes;
         this.freeIndexPointer = freeIndexPointer;
 
         occupied = new BitSet(capacity);
         occupied.set(0, capacity);
 
-        for (int i = 0; i <= freeIndexPointer; i++) {
-            occupied.clear(freeIndexes[i]);
+        for (int i = 0; i < freeIndexPointer; i++) {
+            if (i < freeIndexes.length) {
+                occupied.clear(freeIndexes[i]);
+            } else {
+                occupied.clear(capacity - i - 1);
+            }
+        }
+
+    }
+
+    // the following is only used in testing
+    public IndexManager(int[] freeIndexes, int freeIndexPointer) {
+        this(freeIndexes.length, freeIndexes, freeIndexPointer);
+    }
+
+    public IndexManager(int capacity, int oldCapacity) {
```

Comment:
I don't think this makes sense as a public constructor, the use is very specific to creating a resized Indexmanager.

Reply:
Removed.